Revamp WebCodecs documentation (#43697)
mdn/content: files/en-us/web/api/webcodecs_api/video_processing_concepts/index.md

---
title: Video processing concepts
slug: Web/API/WebCodecs_API/Video_processing_concepts
page-type: guide
---

{{DefaultAPISidebar("WebCodecs API")}}

Before working with the WebCodecs API, it is helpful to understand some foundational concepts around how video works, how it is compressed, and how video files are structured.
This guide covers the key concepts: video frames, codecs, encoding and decoding, containers, and muxing and demuxing.

## Video frames

A video is a sequence of images displayed in rapid succession. Each image in the sequence is called a **video frame**, and each frame has an associated timestamp indicating when it should be displayed.

![Video Frames](video-frames.png)

> Each pixel in a video frame is represented by a set of numeric color channel values. Uncompressed, a single 4K frame (~8 million pixels) is approximately 25 MB. At 30 frames per second, one hour of uncompressed 4K video would be around 750 GB, which is impractically large for storage or streaming.

Reviewer: I don't think  750 GB is the right number here.

Author: Sorry, thanks for catching this. 3840x2160 = ~8.29 million pixels, 4 bytes per pixel => ~33x10⁶ bytes per frame, 30 frames per second => ~1x10⁹  bytes per second, 3600s/h => ~3.6 x10¹² bytes per hour -> ~3.6TB (obviously some rounding errors but close enough?)

![Throwing away high detail information](dct.png)

The amount of detail removed is controlled by two things: the **bitrate**, which determines how much data the output stream uses, and the **codec string**, which specifies the profile and level that govern the encoding logic. Higher bitrates and more capable profiles preserve more detail at the cost of larger file sizes. The following shows the tradeoff between quality and bitrate, using baseline `vp9` on a 1080p video:

![Bitrate ladder](bitrate-ladder.png)

### Temporal compression

Successive frames in a video are typically visually similar to one another. Instead of encoding each video frame as an independent image, video codecs calculate the difference between frames, and encode just the frame differences in a compact binary representation. Codecs typically use a number of techniques such as motion compensation to reduce the amount of data required to encode frame differences.

![Frame differences](frame-diff.png)
Codecs will then store the first video frame in a sequence as a key frame, and then store subsequent frames as just frame differences (called delta frames).

![Key frames vs delta frames](key-frames.png)

> Videos are typically encoded with key frames at regular intervals. To reconstruct a given delta frame, it is necessary to decode the previous key frame, and then all the previous delta frames, in order, up until the current delta frame, in order to properly add up all the frame differences and construct the full current frame for display.

Reviewer 2: ```suggestion
Videos are typically encoded with key frames at regular intervals. To construct the full current frame for display of a given delta frame, you have to decode the previous key frame and all the subsequent delta frames (in order) up to the current delta frame.
```

> In WebCodecs, the `EncodedVideoChunk` interface has a `type` property which can take the value `"key"` or `"delta"` denoting whether or not the chunk represents a key frame or a delta frame.

Reviewer 2: ```suggestion
In WebCodecs, the `EncodedVideoChunk` interface has a `type` property, which can take the value `"key"` or `"delta"` to indicate whether or not the chunk represents a key frame or a delta frame.
```

Because delta frames depend on all previous frames since the last key frame, a decoder cannot start decoding from an arbitrary point in a video — it must always start from a key frame. This has two practical implications: **seeking** to a specific timestamp requires finding the nearest preceding key frame and decoding every frame in order up to the target, and **error recovery** requires skipping forward to the next key frame before resuming decoding.

When encoding with a `VideoEncoder`, it is possible to determine when to set a video as a key frame or a delta frame by using the `keyFrame` parameter in the encoder method.

```js
 encoder.encode(frame, { keyFrame: /* */ })
```

## Encoding and decoding

### Codec compatibility

> For codecs to be useful, it is necessary to be able to both encode video (turn raw video into compressed binary data) with a codec, and to be able to decode the same video (turn the compressed binary data back into raw video frames) with the same codec. The video industry has therefore coalesced around a handful of standard codecs such as `vp9`, `h264`, `hevc` and `av1`.

Reviewer 2: Too many "it is necessary" can be a little irritating - sometimes break it up with "you", or "code has to"
```suggestion
For codecs to be useful, you have to be able to both encode video (turn raw video into compressed binary data) with a codec, and to be able to decode the same video (turn the compressed binary data back into raw video frames) with the same codec. The video industry has therefore coalesced around a handful of standard codecs such as `vp9`, `h264`, `hevc` and `av1`.
```

> Applications which primarily create video content (e.g., video editing tools), and therefore primarily encode video, typically choose a video codec for encoding in order to maximize compatibility with video player software.

Reviewer 2: ```suggestion
Applications that primarily create video content (e.g., video editing tools), and therefore primarily encode video, typically choose a video codec for encoding in order to maximize compatibility with video player software.
```

---
title: Video processing concepts
slug: Web/API/WebCodecs_API/Video_processing_concepts
page-type: guide
---

{{DefaultAPISidebar("WebCodecs API")}}

Before working with the WebCodecs API, it is helpful to understand some foundational concepts around how video works, how it is compressed, and how video files are structured.
This guide covers the key concepts: video frames, codecs, encoding and decoding, containers, and muxing and demuxing.

## Video frames

A video is a sequence of images displayed in rapid succession. Each image in the sequence is called a **video frame**, and each frame has an associated timestamp indicating when it should be displayed.

![Video Frames](video-frames.png)

> Each pixel in the video frame is composed of 3 components, a Red, Green and Blue value (also called an RGB value). Each color channel is represented by 1 byte (a uint8 representing an integer from 0-255, indicating the intensity of the color channel for that pixel), meaning that each pixel is represented by at least 3 bytes of information.

Reviewer 3: This depends on the format. I.e., YUV formats won't be represented this way. I'd probably just avoid talking about the representation here. I'm not sure it's that useful.

Uncompressed, a single 4K frame (~8 million pixels at 3 bytes per pixel) is approximately 25 MB. At 30 frames per second, one hour of uncompressed 4K video would be around 750 GB, which is impractically large for storage or streaming.

Codecs were developed in order to compress video, typically by 1-2 orders of magnitude, to be able to practically store and stream video content given typical device network and storage constraints.

## Codecs

A **codec** (short for encode/decode) is an algorithm for compressing and decompressing video data. Codecs reduce file size dramatically — typically by a factor of 100 or more through a variety of different techniques. While there are a number of video codecs used within the browser, such as `vp9`, `av1` and `h264`, they all apply some form of the following techniques:

### Color space transformations

Codecs will transform the original RGB color values into the YUV colorspace, with the Y channel capturing changes in brightness, and UV channels capturing the other color information. The codecs will then subsample the UV color channels, reducing data use by ~50% for minimal perceived quality loss.

![YUV subsampling](yuv.svg)

> This colorspace transformation shows up within the `format` property of the `VideoFrame` object. The `format` can take the following values:

Reviewer 3: There are a lot more values here, I'd probably link to the pixel format part of the spec.

- `I420` - Normal YUV subsampling, with a Y channel and sub-sampled U and V channels.
- `NV12` - Similar to YUV, but instead of storing data in 3 planes, it interleaves the UV data together into one plane
- `RGBA` - Standard RGB but with an alpha channel
- `BGRA` - Same as RGB, but in reverse order (Blue, Green, Red)

> ### Spatial Compression

Reviewer 3: Dunno if this is useful to detail.
